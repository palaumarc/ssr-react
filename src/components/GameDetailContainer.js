import React from "react";
import { connect } from "react-redux";
import { getGameById, shouldLoadGames, getLastGameRun, shouldLoadGameRuns } from "../reducers";
import moment from "moment";
import { loadGameDetail, resetGameRuns } from "../actions";
import { Link } from "react-router-dom";

import GameDetail from "./GameDetail"

class GameDetailContainer extends React.Component {

    componentDidMount() {
        if (this.props.shouldLoadGames || this.props.shouldLoadGameRuns) {
            this.props.loadGameDetail(this.props.match.params.id);
        }
    }

    componentWillUnmount() {
        this.props.resetGameRuns();
    }

    render() {

        const { lastRun, game } = this.props;

        if (!game || !lastRun) {
            return null;
        }

        const lastRunDuration = moment.duration(lastRun.times.primary);
        const gameName = game.names.international;
        const gameLogoUrl = game.assets["cover-tiny"].uri;
        const lastRunVideoUrl = lastRun.videos.links[0].uri;
        const lastRunPlayerName = lastRun.players[0].name || lastRun.players[0].id;

        const gameInfo = {
            name: gameName,
            logoUrl: gameLogoUrl
        };

        const lastRunInfo = {
            videoUrl: lastRunVideoUrl,
            playerName: lastRunPlayerName,
            duration: lastRunDuration
        }

        return (
            <div>
                <Link to="/"><button>Games</button></Link>
                <GameDetail
                    game={gameInfo}
                    lastRun={lastRunInfo}
                />
            </div>
        );

    }

}

GameDetailContainer.serverFetch = (params) => loadGameDetail(params.id); // static declaration of data requirements

const mapStateToProps = (state, props) => ( {
    game: getGameById(state, props.match.params.id),
    lastRun: getLastGameRun(state),
    shouldLoadGameRuns: shouldLoadGameRuns(state),
    shouldLoadGames: shouldLoadGames(state)
});

const mapDispatchToProps = {
    loadGameDetail,
    resetGameRuns
}

export default connect(mapStateToProps, mapDispatchToProps)(GameDetailContainer);
