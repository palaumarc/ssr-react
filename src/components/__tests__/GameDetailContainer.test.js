import React from "react";
import ReactDOM from "react-dom";
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import renderer from "react-test-renderer";
import GameDetailContainer from "../GameDetailContainer";
import reduxThunk from 'redux-thunk';
import {
    shouldLoadGameRuns,
    shouldLoadGames,
    getGameById,
    getLastGameRun
} from "../../reducers";

import {
    loadGameDetail,
    resetGameRuns
} from "../../actions"

jest.mock("../../reducers");
jest.mock("../../actions");
jest.mock("../GameDetail", () => 'game-detail-mock');

const middlewares = [reduxThunk];
const mockStore = configureStore(middlewares);

//We don't fill store's state as we will mock selectors directly
const state = {};
const store = mockStore(state);

const game = {
    id: '1',
    name: 'dummyName1',
    logoUrl: 'dummyLogoUrl1'
}

const gameRun = {
    videoUrl: 'dummyVideoUrl',
    playerName: 'dummyPlayerName',
    duration: 'PT15M7S'
}

const loadGameDetailMockAction = {type: 'loadGameDetail'};
loadGameDetail.mockReturnValue(loadGameDetailMockAction);
const resetGameRunsMockAction = {type: 'resetGameRuns'};
resetGameRuns.mockReturnValue(resetGameRunsMockAction);

const gameId = game.id;

const componentToTest =
    <Provider store={store}>
        <GameDetailContainer gameId={gameId} />
    </Provider>

beforeEach(() => {
    store.clearActions();
    jest.clearAllMocks();
})

test("Component is unmounted. ResetGameRuns action.", () => {
    shouldLoadGames.mockReturnValue(false);
    shouldLoadGameRuns.mockReturnValue(false);
    getGameById.mockReturnValue(game);
    getLastGameRun.mockReturnValue(gameRun);

    const container = document.createElement('div');
    ReactDOM.render(componentToTest, container);
    ReactDOM.unmountComponentAtNode(container);
    expect(resetGameRuns).toHaveBeenCalledTimes(1);
    expect(resetGameRuns).toHaveBeenCalledWith();
    expect(store.getActions()).toContainEqual(resetGameRunsMockAction);
});

test("ShouldLoadGames returns true. Component is rendered. LoadGameDetail action should be dispatched.", () => {
    shouldLoadGames.mockReturnValue(true);
    shouldLoadGameRuns.mockReturnValue(false);
    getGameById.mockReturnValue(game);
    getLastGameRun.mockReturnValue(gameRun);

    const container = document.createElement('div');

    ReactDOM.render(componentToTest, container);
    expect(loadGameDetail).toHaveBeenCalledTimes(1);
    expect(loadGameDetail).toHaveBeenCalledWith(gameId);
    expect(store.getActions()).toHaveLength(1);
    expect(store.getActions()).toContainEqual(loadGameDetailMockAction);
});

test("ShouldLoadGameRuns returns true. Component is rendered. LoadGameDetail action should be dispatched.", () => {
    shouldLoadGames.mockReturnValue(false);
    shouldLoadGameRuns.mockReturnValue(true);
    getGameById.mockReturnValue(game);
    getLastGameRun.mockReturnValue(gameRun);

    const container = document.createElement('div');

    ReactDOM.render(componentToTest, container);
    expect(loadGameDetail).toHaveBeenCalledTimes(1);
    expect(loadGameDetail).toHaveBeenCalledWith(gameId);
    expect(store.getActions()).toHaveLength(1);
    expect(store.getActions()).toContainEqual(loadGameDetailMockAction);
});

test("ShouldLoadGames and shouldLoadGameRuns returns false. Component is rendered. LoadGameDetail action should not be dispatched.", () => {
    shouldLoadGames.mockReturnValue(false);
    shouldLoadGameRuns.mockReturnValue(false);
    getGameById.mockReturnValue(game);
    getLastGameRun.mockReturnValue(gameRun);

    const container = document.createElement('div');

    ReactDOM.render(componentToTest, container);
    expect(loadGameDetail).not.toHaveBeenCalled();
    expect(store.getActions()).toHaveLength(0);
});

test("GetGameById returns undefined. Snapshot.", () => {
    shouldLoadGames.mockReturnValue(false);
    shouldLoadGameRuns.mockReturnValue(false);
    getGameById.mockReturnValue(undefined);
    getLastGameRun.mockReturnValue(gameRun);

    const tree = renderer.create(componentToTest).toJSON();
    expect(tree).toMatchSnapshot();
});

test("GetLastGameRun returns undefined. Snapshot.", () => {
    shouldLoadGames.mockReturnValue(false);
    shouldLoadGameRuns.mockReturnValue(false);
    getGameById.mockReturnValue(game);
    getLastGameRun.mockReturnValue(undefined);

    const tree = renderer.create(componentToTest).toJSON();
    expect(tree).toMatchSnapshot();
});

test("GetGameById and getLastGameRun return valid values. Snapshot.", () => {
    shouldLoadGames.mockReturnValue(false);
    shouldLoadGameRuns.mockReturnValue(false);
    getGameById.mockReturnValue(game);
    getLastGameRun.mockReturnValue(gameRun);

    const tree = renderer.create(componentToTest).toJSON();
    expect(tree).toMatchSnapshot();
});

