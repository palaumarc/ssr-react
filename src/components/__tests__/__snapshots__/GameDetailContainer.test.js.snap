// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetGameById and getLastGameRun return valid values. Snapshot. 1`] = `
<div>
  <link-mock
    to="/"
  >
    <button
      style={
        Object {
          "backgroundColor": "#008CBA",
          "border": "none",
          "borderRadius": "4px",
          "color": "white",
          "cursor": "pointer",
          "fontSize": "20px",
          "marginBottom": "20px",
          "textAlign": "center",
        }
      }
    >
      Games
    </button>
  </link-mock>
  <game-detail-mock
    game={
      Object {
        "logoUrl": "dummyLogoUrl1",
        "name": "dummyName1",
      }
    }
    lastRun={
      Object {
        "duration": "PT15M7S",
        "playerName": "dummyPlayerName",
        "videoUrl": "dummyVideoUrl",
      }
    }
  />
</div>
`;

exports[`GetGameById returns undefined. Snapshot. 1`] = `
<div>
  Game not found!
</div>
`;

exports[`GetLastGameRun returns undefined. Snapshot. 1`] = `
<div>
  No runs available!
</div>
`;
